compiler/semstmts.nim

```diff
@@ -445,55 +449,114 @@ proc setVarType(c: PContext; v: PSym, typ: PType) =
-        return nil
-      sym = searchInScopes(c, ident, amb)
-      # XXX what if amb is true?
-      # CHECKME: should that test also apply to `nkSym` case?
-      if sym == nil or sfCustomPragma in sym.flags: return nil
-    of nkSym:
-      sym = nodePragma.sym
-    else:
-      return nil
-      # skip if not in scope; skip `template myAttr() {.pragma.}`
-
-    let lhs = b[0]
-    let clash = strTableGet(c.currentScope.symbols, lhs.ident)
-    if clash != nil:
-      # refs https://github.com/nim-lang/Nim/issues/8275
-      wrongRedefinition(c, lhs.info, lhs.ident.s, clash.info)
-
-    result = newTree(nkCall)
-    result.add nodePragma
-    result.add lhs
-    if a[1].kind != nkEmpty:
-      result.add a[1]
-    else:
-      result.add newNodeIT(nkNilLit, a.info, c.graph.sysTypes[tyNil])
-    result.add a[2]
-    result.info = a.info
-    let ret = newNodeI(nkStmtList, a.info)
-    ret.add result
-    result = semExprNoType(c, ret)
+    const
+      namePos = 0
+      pragmaPos = 1
+    let pragmas = b[pragmaPos]
+    for i in 0 ..< pragmas.len:
+      # Mirrored with semProcAnnotation
+
+      let it = pragmas[i]
+      let key = if it.kind in nkPragmaCallKinds and it.len >= 1: it[0] else: it
+      
+      when false:
+        let lhs = b[0]
+        let clash = strTableGet(c.currentScope.symbols, lhs.ident)
+        if clash != nil:
+          # refs https://github.com/nim-lang/Nim/issues/8275
+          wrongRedefinition(c, lhs.info, lhs.ident.s, clash.info)
+
+      if whichPragma(it) != wInvalid or key.kind notin nkIdentKinds:
+        # Not a custom pragma
+        continue
+      else:
+        let ident = considerQuotedIdent(c, key)
+        if strTableGet(c.userPragmas, ident) != nil:
+          continue # User defined pragma
+        else:
+          var amb = false
+          let sym = searchInScopes(c, ident, amb)
+          if sym != nil and sfCustomPragma in sym.flags:
+            continue # User custom pragma
+
+      # we transform ``proc p {.m, rest.}`` into ``m(do: proc p {.rest.})`` and
+      # let the semantic checker deal with it:
+      var x = newNodeI(nkCall, key.info)
+      x.add(key)
+
+      if it.kind in nkPragmaCallKinds and it.len > 1:
+        # pass pragma arguments to the macro too:
+        for i in 1..<it.len:
+          x.add(it[i])
+
+      # Drop the pragma from the list, this prevents getting caught in endless
+      # recursion when the nkCall is semanticized
+      let oldExpr = a[lhsPos]
+      let newPragmas = copyExcept(pragmas, i)
+      if newPragmas.kind != nkEmpty and newPragmas.len == 0:
+        a[lhsPos] = oldExpr[namePos]
+      else:
+        a[lhsPos] = copyNode(oldExpr)
+        a[lhsPos].add(oldExpr[namePos])
+        a[lhsPos].add(newPragmas)
+
+      var unarySection = newNodeI(n.kind, a.info)
+      unarySection.add(a)
+      x.add(unarySection)
+
+      # recursion assures that this works for multiple macro annotations too:
```

Comment:
I know it's done this way in semProcAux too, but's too subtle. There must be a cleaner way of accomplishing the same.

Reply:
It should be documented as "only the first macro pragma from left to right will be evaluated, this macro can choose whether or not to return a proc with the remaining macro pragmas". That sounds bad but I don't see any way that makes more sense.